File: software/examples/java/ExampleCallback.java
import com.tinkerforge.IPConnection;
import com.tinkerforge.BrickIMU;

public class ExampleCallback {
	private static final String HOST = "localhost";
	private static final int PORT = 4223;

	// Change XXYYZZ to the UID of your IMU Brick
	private static final String UID = "XXYYZZ";

	// Note: To make the example code cleaner we do not handle exceptions. Exceptions
	//       you might normally want to catch are described in the documentation
	public static void main(String args[]) throws Exception {
		IPConnection ipcon = new IPConnection(); // Create IP connection
		BrickIMU imu = new BrickIMU(UID, ipcon); // Create device object

		ipcon.connect(HOST, PORT); // Connect to brickd
		// Don't use device before ipcon is connected

		// Add quaternion listener
		imu.addQuaternionListener(new BrickIMU.QuaternionListener() {
			public void quaternion(float x, float y, float z, float w) {
				System.out.println("Quaternion[X]: " + x);
				System.out.println("Quaternion[Y]: " + y);
				System.out.println("Quaternion[Z]: " + z);
				System.out.println("Quaternion[W]: " + w);
				System.out.println("");
			}
		});

		// Set period for quaternion callback to 1s (1000ms)
		imu.setQuaternionPeriod(1000);

		System.out.println("Press key to exit"); System.in.read();
		ipcon.disconnect();
	}
}

File: software/examples/c/example_callback.c
#include <stdio.h>

#include "ip_connection.h"
#include "brick_imu.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XXYYZZ" // Change XXYYZZ to the UID of your IMU Brick

// Callback function for quaternion callback
void cb_quaternion(float x, float y, float z, float w, void *user_data) {
	(void)user_data; // avoid unused parameter warning

	printf("Quaternion [X]: %f\n", x);
	printf("Quaternion [Y]: %f\n", y);
	printf("Quaternion [Z]: %f\n", z);
	printf("Quaternion [W]: %f\n", w);
	printf("\n");
}

int main(void) {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	IMU imu;
	imu_create(&imu, UID, &ipcon);

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		return 1;
	}
	// Don't use device before ipcon is connected

	// Register quaternion callback to function cb_quaternion
	imu_register_callback(&imu,
	                      IMU_CALLBACK_QUATERNION,
	                      (void (*)(void))cb_quaternion,
	                      NULL);

	// Set period for quaternion callback to 1s (1000ms)
	imu_set_quaternion_period(&imu, 1000);

	printf("Press key to exit\n");
	getchar();
	imu_destroy(&imu);
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
	return 0;
}
